[Filters] Support CoreGraphics filters

shallawa opened the pull request:
#### f13138563688bcbd1a47b72e4ffdb98dd4ba9ffe
<pre>
[Filters] Support CoreGraphics filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=243816">https://bugs.webkit.org/show_bug.cgi?id=243816</a>

Reviewed by NOBODY (OOPS!).

CoreGraphics provides a better way to apply FilterEffects. A CGStyle can be used
to apply the a FilterEffect to the current drawing. Also beginTransparencyLayer()
/endTransparencyLayer() can be used to force applying the FilterEffect only when
composting the layer to the destination GraphicsContext. This will treat the drawing
commands between beginning and ending the transparency layer as a unit of drawing.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFilterImageValue.cpp:
(WebCore::CSSFilterImageValue::image):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterMode const):
(WebCore::Page::acceleratedFiltersEnabled const): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::addFilterEffectStyles):
(WebCore::GraphicsContext::removeFilterEffectStyles):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::filterEffectStyle const):
(WebCore::GraphicsContext::setFilterEffectStyle):
* Source/WebCore/platform/graphics/GraphicsContextState.cpp:
(WebCore::GraphicsContextState::mergeChanges):
(WebCore::stateChangeName):
(WebCore::GraphicsContextState::dump const):
* Source/WebCore/platform/graphics/GraphicsContextState.h:
(WebCore::GraphicsContextState::filterEffectStyle const):
(WebCore::GraphicsContextState::setFilterEffectStyle):
(WebCore::GraphicsContextState::encode const):
(WebCore::GraphicsContextState::decode):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGFilterEffectStyle):
(WebCore::GraphicsContextCG::didUpdateState):
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterModes const):
(WebCore::FEColorMatrix::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::supportedFilterModes const):
(WebCore::FEComponentTransfer::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
* Source/WebCore/platform/graphics/filters/FEDropShadow.cpp:
(WebCore::FEDropShadow::supportedFilterModes const):
(WebCore::FEDropShadow::createStyle const):
* Source/WebCore/platform/graphics/filters/FEDropShadow.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::Filter):
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::filterMode const):
(WebCore::Filter::setFilterMode):
(WebCore::Filter::renderingMode const):
(WebCore::Filter::Filter):
(WebCore::Filter::setRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.h:
(WebCore::FilterEffect::createStyle const):
* Source/WebCore/platform/graphics/filters/FilterEffectStyle.h: Added.
(WebCore::operator==):
(WebCore::FEDropShadowStyle::encode const):
(WebCore::FEDropShadowStyle::decode):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
(WebCore::FilterFunction::supportedFilterModes const):
(WebCore::FilterFunction::createStyles const):
(WebCore::FilterFunction::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterMode.h: Added.
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterModes const):
(WebCore::SourceGraphic::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/SourceGraphic.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
(WebCore::CSSFilter::CSSFilter):
(WebCore::createReferenceFilter):
(WebCore::CSSFilter::supportedFilterModes const):
(WebCore::CSSFilter::createStyles const):
(WebCore::CSSFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
(WebCore::RenderLayerFilters::applyFilterEffect):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
(WebCore::RenderSVGResourceFilter::postApplyResource):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
(WebCore::SVGFilter::supportedFilterModes const):
(WebCore::SVGFilter::createStyles const):
(WebCore::SVGFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/Platform/IPC/FilterReference.h:
(IPC::FilterReference::encodeFilter):
(IPC::FilterReference::decodeFilter):
</pre>

smfr commented on Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
```diff
@@ -663,6 +663,21 @@ MuteCameraOnMicrophoneInterruptionEnabled:
     WebCore:
       default: false
 
+NativeFiltersEnabled:
```
Not really a fan of "Native" - it doesn't really say anything. Maybe "GraphicsContext Filters" or something?

shallawa replied:
Done. This was changed to `GraphicsContextFiltersEnabled`.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3759,13 +3759,17 @@ void Page::recomputeTextAutoSizingInAllFrames()
 #endif
 
-bool Page::acceleratedFiltersEnabled() const
+FilterMode Page::preferredFilterMode() const
```
I guess this is OK for now, but at some point we might want to have both "native" and Core Image filters enabled, using native when possible, falling back to Core Image, and then falling back to software, depending on the filter graph.

smfr commented on Source/WebCore/platform/graphics/GraphicsContext.cpp:
```diff
@@ -81,6 +81,23 @@ void GraphicsContext::restore()
+void GraphicsContext::addFilterEffectStyles(const Vector<FilterEffectStyle> styles)
+{
+    for (auto it = styles.rbegin(); it != styles.rend(); ++it) {
+        save();
```
I don't think the `save()` and the `beginTransparencyLayer()` should live here in `GraphicsContext`. It should be the responsibility of the caller to do those things. Setting a filter style should behave exactly like setting other state like fillColor.

shallawa replied:
Done. This function is moved to Filter::applyStyles().

smfr commented on Source/WebCore/platform/graphics/GraphicsContext.cpp:
```diff
@@ -81,6 +81,23 @@ void GraphicsContext::restore()
+    }
+}
+
+void GraphicsContext::removeFilterEffectStyles(unsigned count)
```
That means there is no need for a "remove" function.

shallawa replied:
Done. This function is moved to Filter::removeStyles().

smfr commented on Source/WebCore/platform/graphics/GraphicsContextState.cpp:
```diff
@@ -193,6 +219,7 @@ TextStream& GraphicsContextState::dump(TextStream& ts) const
 
     dump(Change::CompositeMode,                 &GraphicsContextState::m_compositeMode);
     dump(Change::DropShadow,                    &GraphicsContextState::m_dropShadow);
+//    dump(Change::FilterEffectStyle,             &GraphicsContextState::m_filterEffectStyle);
```
Please write the dumping code!

shallawa replied:
Done.

smfr commented on Source/WebCore/platform/graphics/GraphicsContextState.h:
```diff
@@ -97,6 +99,9 @@ class GraphicsContextState {
     const DropShadow& dropShadow() const { return m_dropShadow; }
     void setDropShadow(const DropShadow& dropShadow) { setProperty(Change::DropShadow, &GraphicsContextState::m_dropShadow, dropShadow); }
 
+    const std::optional<FilterEffectStyle>& filterEffectStyle() const { return m_filterEffectStyle; }
```
Won't this subsume dropShadow()?

shallawa replied:
Yes. It can be done in a separate cleaning patch.

smfr commented on Source/WebCore/platform/graphics/filters/FilterEffectStyle.h:
```diff
@@ -0,0 +1,80 @@
+
+namespace WebCore {
+
+struct FEDropShadowStyle {
```
Maybe `FilterStyleDropShadow`, then we can add `FilterStyleBlur`, `FilterStyleColorMatrix`.

shallawa replied:
I changed the code a little. This struct is now called `GraphicsStyleDropShadow`. And the variant of all the styles is now called `GraphicsStyle`. I think it is better to omit the prefix 'Filter' from the GraphicsContextState member since CoreGraphics calls it CGStyle. I used FilterStyle for another struct which will include the filter primitive geometry.

smfr commented on Source/WebCore/rendering/RenderLayerFilters.cpp:
```diff
@@ -160,7 +160,7 @@ GraphicsContext* RenderLayerFilters::beginFilterEffect(RenderElement& renderer,
         m_targetBoundingBox = targetBoundingBox;
         // FIXME: This rebuilds the entire effects chain even if the filter style didn't change.
-        m_filter = CSSFilter::create(renderer, renderer.style().filter(), m_renderingMode, m_filterScale, Filter::ClipOperation::Unite, m_targetBoundingBox, context);
+        m_filter = CSSFilter::create(renderer, renderer.style().filter(), m_filterMode, m_filterScale, Filter::ClipOperation::Unite, m_targetBoundingBox, context);
```
The `mode` argument to `CSSFilter::create` should be a `preferredMode`, right? Somewhere there should be some code that loops over the list of filters and computes which mode can be used for that set of filters. Then that's the mode that actually gets used.

shallawa replied:
This happened in `CSSFilter::create()` after building all the FilterFunctions by this statement:

```
    if (filterMode != FilterMode::Software && !filter->supportedFilterModes().contains(filterMode))
        filter->setFilterMode(FilterMode::Software);

```

smfr commented on Source/WebCore/rendering/RenderLayerFilters.cpp:
```diff
@@ -199,6 +199,13 @@ GraphicsContext* RenderLayerFilters::beginFilterEffect(RenderElement& renderer,
+
+    if (filter.filterMode() == FilterMode::Native) {
+        auto styles = filter.createStyles();
+        context.addFilterEffectStyles(styles);
```
This is where you need to save the state, set the style and start a transparency layer for each filter.

smfr commented on Source/WebCore/rendering/RenderLayerFilters.cpp:
```diff
@@ -218,11 +225,16 @@ void RenderLayerFilters::applyFilterEffect(GraphicsContext& destinationContext)
-    destinationContext.drawFilteredImageBuffer(m_sourceImage.get(), m_filterRegion, *m_filter, results);
+    if (m_filter->filterMode() == FilterMode::Native) {
+        auto styles = m_filter->createStyles();
+        destinationContext.removeFilterEffectStyles(styles.size());
```
This will be a set of `endTransparencyLayer` and `restore`. It's unfortunate that we have to create the styles against; this seems error prone.

smfr commented on Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
```diff
@@ -87,6 +87,11 @@ bool RenderSVGResourceFilter::applyResource(RenderElement& renderer, const Rende
         FilterData* filterData = m_rendererFilterDataMap.get(&renderer);
+        if (!filterData->sourceImage) {
+            auto styles = filterData->filter->createStyles();
+            context->addFilterEffectStyles(styles);
```
This will change too. But why is the code not consulting the mode?

shallawa replied:
Done.

smfr commented on Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
```diff
@@ -203,7 +212,12 @@ void RenderSVGResourceFilter::postApplyResource(RenderElement& renderer, Graphic
+        if (filterData.sourceImage)
+            context->drawFilteredImageBuffer(filterData.sourceImage.get(), filterData.sourceImageRect, *filterData.filter, filterData.results);
+        else {
+            auto styles = filterData.filter->createStyles();
```
Rather than assume null sourceImage mean "native", it would be clearer to test the mode here.

shallawa replied:
Done. You are right.

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -235,6 +221,16 @@ FilterEffectVector SVGFilter::effectsOfType(FilterFunction::Type filterType) con
+    OptionSet<FilterMode> modes;
+
+    for (auto& term : m_expression)
+        modes.add(term.effect->supportedFilterModes());
```
Isn't the result mode the intersection of the supported modes for each effect, not their union?

shallawa replied:
Yes you are right. Fixed.

smfr commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -277,6 +273,33 @@ RefPtr<FilterImage> SVGFilter::apply(FilterImage* sourceImage, FilterResults& re
+    return createStyles();
+}
+
+Vector<FilterEffectStyle> SVGFilter::createStyles() const
```
How does this work when the graph is not just a list? We can't render non-linear graphs with native filters.

shallawa replied:
This function should not be reached. The FilterMode::GraphicsContext will be added by the FilterEffects which take only one input. I added the following assertion to this function:

```
ASSERT(supportedFilterModes().contains(FilterMode::GraphicsContext));
```

weinig commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -1073,6 +1073,23 @@ static void setCGShadow(CGContextRef context, RenderingMode renderingMode, const
         CGContextSetShadowWithColor(context, CGSizeMake(xOffset, yOffset), blurRadius, cachedCGColor(color).get());
 }
 
+#if PLATFORM(COCOA)
```
I generally prefer for these to be a bit more narrowly scoped when we can, and I think we can in this case. Maybe HAVE(CORE_GRAPHICS_CREATE_SHADOW2) or something to that effect?

shallawa replied:
Done. I used `HAVE(CORE_GRAPHICS_FILTERS)` to gird this code. If more CoreGraphics styles are added in the future, new HAVE macros will be added.

weinig commented on Source/WebCore/platform/graphics/filters/FEDropShadow.cpp:
```diff
@@ -126,6 +126,25 @@ std::unique_ptr<FilterEffectApplier> FEDropShadow::createSoftwareApplier() const
+OptionSet<FilterMode> FEDropShadow::supportedFilterModes() const
+{
+    return {
+#if PLATFORM(COCOA)
```
Again, wherever possible, not using PLATFORM() macros, but instead using scoped ones is preferred as it is better self documenting and provides an easy way to search for places where it is needed.

shallawa replied:
Done.

weinig commented on Source/WebCore/platform/graphics/filters/Filter.cpp:
```diff
@@ -102,4 +103,21 @@ RefPtr<FilterImage> Filter::apply(ImageBuffer* sourceImage, const FloatRect& sou
+void Filter::applyStyles(GraphicsContext& context) const
+{
+    for (auto& style : createStyles()) {
+        context.save();
```
Man these disconnected save and restores make me nervous. I wonder if there is a way we can ensure these are always both invoked statically?

shallawa replied:
I think we can do something like this:

```
void GraphicsContext::drawFilteredImage(const Filter& filter, Function<void (GraphicsContext&)>&& drawCallback)
{
    filter.applyStyles(*this);
    drawCallback(*this);
    filter.removeStyles(*this);
}

```

Or

```
void Filter::drawFilteredImage(GraphicsContext& context, Function<void (GraphicsContext&)>&& drawCallback)
{
    auto filterStyles = createFilterStyles(sourceImageRect);

    for (auto& filterStyle : filterStyles) {
        context.save();
        context.clip(filterStyle.imageRect);
        context.setStyle(filterStyle.style);
        context.beginTransparencyLayer(1);
    }

    drawCallback(context);

    for (unsigned i = 0; i < filterStyles.size(); --i) {
        context.endTransparencyLayer();
        context.restore();
    }
}

```

shallawa committed:
[Filters] Support CoreGraphics filters
https://bugs.webkit.org/show_bug.cgi?id=243816

Reviewed by NOBODY (OOPS!).

CoreGraphics provides a better way to apply FilterEffects. A CGStyle can be used
to apply the a FilterEffect to the current drawing. Also beginTransparencyLayer()
/endTransparencyLayer() can be used to force applying the FilterEffect only when
composting the layer to the destination GraphicsContext. This will treat the drawing
commands between beginning and ending the transparency layer as a unit of drawing.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSFilterImageValue.cpp:
(WebCore::CSSFilterImageValue::image):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterMode const):
(WebCore::Page::acceleratedFiltersEnabled const): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::addFilterEffectStyles):
(WebCore::GraphicsContext::removeFilterEffectStyles):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::filterEffectStyle const):
(WebCore::GraphicsContext::setFilterEffectStyle):
* Source/WebCore/platform/graphics/GraphicsContextState.cpp:
(WebCore::GraphicsContextState::mergeChanges):
(WebCore::stateChangeName):
(WebCore::GraphicsContextState::dump const):
* Source/WebCore/platform/graphics/GraphicsContextState.h:
(WebCore::GraphicsContextState::filterEffectStyle const):
(WebCore::GraphicsContextState::setFilterEffectStyle):
(WebCore::GraphicsContextState::encode const):
(WebCore::GraphicsContextState::decode):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGFilterEffectStyle):
(WebCore::GraphicsContextCG::didUpdateState):
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterModes const):
(WebCore::FEColorMatrix::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::supportedFilterModes const):
(WebCore::FEComponentTransfer::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
* Source/WebCore/platform/graphics/filters/FEDropShadow.cpp:
(WebCore::FEDropShadow::supportedFilterModes const):
(WebCore::FEDropShadow::createStyle const):
* Source/WebCore/platform/graphics/filters/FEDropShadow.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::Filter):
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::filterMode const):
(WebCore::Filter::setFilterMode):
(WebCore::Filter::renderingMode const):
(WebCore::Filter::Filter):
(WebCore::Filter::setRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.h:
(WebCore::FilterEffect::createStyle const):
* Source/WebCore/platform/graphics/filters/FilterEffectStyle.h: Added.
(WebCore::operator==):
(WebCore::FEDropShadowStyle::encode const):
(WebCore::FEDropShadowStyle::decode):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
(WebCore::FilterFunction::supportedFilterModes const):
(WebCore::FilterFunction::createStyles const):
(WebCore::FilterFunction::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterMode.h: Added.
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterModes const):
(WebCore::SourceGraphic::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/SourceGraphic.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
(WebCore::CSSFilter::CSSFilter):
(WebCore::createReferenceFilter):
(WebCore::CSSFilter::supportedFilterModes const):
(WebCore::CSSFilter::createStyles const):
(WebCore::CSSFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
(WebCore::RenderLayerFilters::applyFilterEffect):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
(WebCore::RenderSVGResourceFilter::postApplyResource):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
(WebCore::SVGFilter::supportedFilterModes const):
(WebCore::SVGFilter::createStyles const):
(WebCore::SVGFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/Platform/IPC/FilterReference.h:
(IPC::FilterReference::encodeFilter):
(IPC::FilterReference::decodeFilter):
